### PR TITLE
Handle JSONDecodeError in Distributed API class methods

### LINKED ISSUE #10212: Unhandled JSONDecodeError in distributed API requests
|Wazuh version|Component|
|---|---|
| 4.3.0 | Distributed API |

## Description

I have encountered the following error during a cluster workload test:
```
2021/09/20 11:12:06 INFO: HTTPS is enabled but cannot find the private key and/or certificate. Attempting to generate them
2021/09/20 11:12:06 INFO: Generated private key file in WAZUH_PATH/api/configuration/ssl/server.key
2021/09/20 11:12:06 INFO: Generated certificate file in WAZUH_PATH/api/configuration/ssl/server.crt
2021/09/20 11:12:13 INFO: Listening on 0.0.0.0:55000..
2021/09/20 11:18:18 INFO: Listening on 0.0.0.0:55000..
2021/09/20 11:35:36 INFO: wazuh 172.31.32.158 "GET /security/user/authenticate" with parameters {} and body {} done in 0.546s: 200
2021/09/20 11:36:10 ERROR: Unhandled exception: Expecting value: line 1 column 1 (char 0)
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 157, in distribute_function
    response = await self.forward_request()
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 514, in forward_request
    response = await asyncio.shield(asyncio.gather(*[forward(node) for node in cleaned_valid_nodes]))
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 423, in forward
    result = json.loads(await client.execute(b'dapi_fwd',
  File "/var/ossec/framework/python/lib/python3.9/json/__init__.py", line 359, in loads
    return cls(**kw).decode(s)
  File "/var/ossec/framework/python/lib/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/var/ossec/framework/python/lib/python3.9/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
2021/09/20 11:36:10 INFO: wazuh 172.31.32.158 "PUT /security/config" with parameters {} and body {"auth_token_exp_timeout": 999999} done in 34.930s: 500
```

This same error has already been corrected in other parts of the code, in this PR: https://github.com/wazuh/wazuh/pull/8901. However, it seems not to be correctly handled when getting it as a response to a distributed call, here:
https://github.com/wazuh/wazuh/blob/bb1449a3bd89644ad8aed335439ceb6f94061555/framework/wazuh/core/cluster/dapi/dapi.py#L419-L434

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
| #10212 |

This PR closes #10212.

As it was commented in the issue, this error can also appear in the `get_nodes` function in `control.py` and it is unhandled too. This would affect the endpoint `GET /cluster/nodes`. It can appear in the rest of functions in `control.py` but `get_system_nodes`. Because of this, it affects `GET /cluster/healthcheck`.

These requests are `local_master`, so the exception is caught in the `execute_local_request` function and not outside (`distribute_function`).

To sum up, we have 2 situations where the `JSONDecodeError` can be caught: **forward_request and execute_remote_request**; and **execute_local_request**.

In both cases the `JSONDecodeError` is caught as a generic `Exception` and a `WazuhInternalError` is generated with code **1000** (unhandled exception).

### Solution

To handle these exceptions, I have added a try except block in `distribute_function` to raise a **WazuhInternalError with code 3036** when we have a JSONDecodeError in `execute:remote_request` or `forward_request` (problem reported in the issue) .

In `execute_local_request`, I have added a new try except block to catch the JSONDecodeError when using `run_local` or `run_local` in the `threadpool`. Again **WazuhInternalError with code 3036**. In this case, we cannot catch the error at a higher level because we catch a generic exception inside the `execute_local_request` function.

#### Examples:

- Remote request (local_master from a worker's API):

```
root@wazuh-worker1:/# curl -k -X GET "https://localhost:55000/cluster/healthcheck" -H "Authorization: Bearer $TOKEN"
{"title": "Wazuh Internal Error", "detail": "JSON couldn't be loaded", "dapi_errors": {"master-node": {"error": "JSON couldn't be loaded", "logfile": "WAZUH_HOME/logs/cluster.log"}}, "error": 3036}
```

worker's api.log:

`2021/09/29 08:35:47 INFO: wazuh 127.0.0.1 "GET /cluster/healthcheck" with parameters {} and body {} done in 0.095s: 500
`

master's cluster.log:

`2021/09/29 08:35:47 ERROR: [Local Client] [D API] JSON couldn't be loaded
`

- Local request (local_master or local_any):

```
root@wazuh-master:/# curl -k -X GET "https://localhost:55000/cluster/healthcheck" -H "Authorization: Bearer $TOKEN"
{"title": "Wazuh Internal Error", "detail": "JSON couldn't be loaded", "dapi_errors": {"master-node": {"error": "JSON couldn't be loaded", "logfile": "WAZUH_HOME/logs/api.log"}}, "error": 3036}
```

master's api.log:

```
2021/09/29 08:33:09 ERROR: JSON couldn't be loaded
2021/09/29 08:33:09 INFO: wazuh 127.0.0.1 "GET /cluster/healthcheck" with parameters {} and body {} done in 0.062s: 500
```

- Forward request (distributed_master):

```
root@wazuh-master:/# curl -k -X PUT "https://localhost:55000/security/user/revoke" -H "Authorization: Bearer $TOKEN"
{"title": "Wazuh Internal Error", "detail": "JSON couldn't be loaded", "dapi_errors": {"master-node": {"error": "JSON couldn't be loaded", "logfile": "WAZUH_HOME/logs/api.log"}}, "error": 3036}
```

master's api.log:

```
2021/09/29 08:38:56 ERROR: JSON couldn't be loaded
2021/09/29 08:38:56 INFO: wazuh 127.0.0.1 "PUT /security/user/revoke" with parameters {} and body {} done in 0.082s: 500
```

### Test results

```
test_agent_DELETE_endpoints.tavern.yaml 
	 6 passed, 8 warnings

test_agent_GET_endpoints.tavern.yaml 
	 91 passed, 93 warnings

test_agent_POST_endpoints.tavern.yaml 
	 5 passed, 7 warnings

test_agent_PUT_endpoints.tavern.yaml 
	 8 passed, 2 xpassed, 12 warnings

test_security_DELETE_endpoints.tavern.yaml 
	 15 passed, 17 warnings

test_security_GET_endpoints.tavern.yaml 
	 11 passed, 13 warnings

test_security_POST_endpoints.tavern.yaml 
	 7 passed, 9 warnings

test_security_PUT_endpoints.tavern.yaml 
	 9 passed, 11 warnings

test_cluster_endpoints.tavern.yaml 
	 45 passed, 47 warnings

test_experimental_endpoints.tavern.yaml 
	 11 passed, 1 xpassed, 14 warnings
```

```
pytest framework/wazuh/core/cluster/dapi/tests/ --disable-warnings 
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/manuel/git/wazuh/framework
plugins: html-3.1.1, metadata-1.11.0, cov-2.12.0, testinfra-5.0.0, asyncio-0.14.0
collected 25 items                                                             

framework/wazuh/core/cluster/dapi/tests/test_dapi.py ................... [ 76%]
......                                                                   [100%]

======================== 25 passed, 9 warnings in 0.60s ========================
```

### CODE REVIEW

File: framework/wazuh/core/cluster/dapi/dapi.py [hunk 1]
```diff
@@ -271,8 +275,8 @@ def run_local():
             return json.dumps(e, cls=c_common.WazuhJSONEncoder)
         except exception.WazuhInternalError as e:
             e.dapi_errors = self.get_error_info(e)
-            # Avoid exception info if it is an asyncio timeout
-            self.logger.error(f"{e.message}", exc_info=True if e.code != 3021 else False)
+            # Avoid exception info if it is an asyncio timeout or JSONDecodeError
+            self.logger.error(f"{e.message}", exc_info=True if e.code not in [3021, 3036] else False)
```
Comment (vicferpoy) [inline framework/wazuh/core/cluster/dapi/dapi.py:279]:
```suggestion
            self.logger.error(f"{e.message}", exc_info=e.code not in {3021, 3036})
```

**Review by vicferpoy — CHANGES_REQUESTED** [review]
These try/except blocks could be unified on a higher level, `distribute_function`, to handle most of the cases.

Missing a new issue to investigate this error as well.

File: framework/wazuh/core/cluster/dapi/dapi.py (hunk 1 above)
Comment (mcarmona99) [reply framework/wazuh/core/cluster/dapi/dapi.py:279]: Adding **suggestion + change of try/except to a higher level** in a single commit

**Review by vicferpoy — APPROVED** [review]
LGTM!

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**